tools/llzk-tblgen/CommonAttrOrTypeCAPIGen.h

```diff
@@ -0,0 +1,341 @@
+      if (isArrayRefType(cppType)) {
+        // For ArrayRef parameters, use intptr_t for count and pointer to element type
+        mlir::StringRef cppElemType = extractArrayRefElementType(cppType);
+        paramListStream << ", intptr_t " << param.getName() << "Count, "
+                        << mapCppTypeToCapiType(cppElemType) << " *" << param.getName();
+      } else {
+        paramListStream << ", " << mapCppTypeToCapiType(cppType) << " " << param.getName();
+      }
+    }
+
+    os << llvm::formatv(
+        fmt,
+        FunctionPrefix,            // {0}
+        kind,                      // {1}
+        dialectNameCapitalized,    // {2}
+        className,                 // {3}
+        paramListBuffer,           // {4}
+        dialect->getCppNamespace() // {5}
+    );
+  }
+
+  void genCompleteRecord(const mlir::tblgen::AttrOrTypeDef def) {
+    const mlir::tblgen::Dialect &defDialect = def.getDialect();
+
+    // Generate for the selected dialect only
+    if (defDialect.getName() != DialectName) {
+      return;
+    }
+
+    this->setDialectAndClassName(&defDialect, def.getCppClassName());
+
+    // Generate IsA check
+    if (GenIsA) {
+      this->genIsADecl();
+    }
+
+    // Generate default Get builder if not skipped
+    if (!def.skipDefaultBuilders()) {
+      this->genDefaultGetBuilderDecl(def);
+    }
+
+    // Generate parameter getters
+    if (GenTypeOrAttrParamGetters) {
+      for (const auto &param : def.getParameters()) {
+        this->setParamName(param.getName());
+        mlir::StringRef cppType = param.getCppType();
+        if (isArrayRefType(cppType)) {
+          this->genArrayRefParameterGetterDecls(cppType);
+        } else {
+          this->genParameterGetterDecl(cppType);
+        }
+      }
+    }
+
+    // Generate extra class method declarations
+    if (GenExtraClassMethods) {
+      std::optional<mlir::StringRef> extraDecls = def.getExtraDecls();
+      if (extraDecls.has_value()) {
+        this->genExtraMethods(extraDecls.value());
+      }
+    }
+  }
+
+protected:
+  mlir::StringRef paramName;
+  std::string paramNameCapitalized;
+};
+
+/// @brief Generator for attribute/type C implementation files
+///
+/// This class extends ImplementationGenerator to provide attribute and type-specific
+/// implementation generation capabilities, including parameter getters and builders.
+struct AttrOrTypeImplementationGenerator : public ImplementationGenerator {
+  using ImplementationGenerator::ImplementationGenerator;
+  virtual ~AttrOrTypeImplementationGenerator() = default;
+
+  /// @brief Set the parameter name for code generation
+  /// @param name The parameter name from the TableGen definition
+  void setParamName(mlir::StringRef name) {
+    this->paramName = name;
+    this->paramNameCapitalized = toPascalCase(name);
+  }
+
+  virtual void genPrologue() const {
+    os << R"(
+#include "mlir/CAPI/IR.h"
+#include "mlir/CAPI/Support.h"
+#include "llvm/ADT/TypeSwitch.h"
```

Comment:
```suggestion
#include <mlir/CAPI/IR.h>
#include <mlir/CAPI/Support.h>
#include <llvm/ADT/TypeSwitch.h>
```
Just a nit, unless this wouldn't work for some reason.